File: java/src/main/java/org/mydotey/quantileestimator/decorator/ValidationDecorator.java
package org.mydotey.quantileestimator.decorator;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import org.mydotey.quantileestimator.QuantileEstimator;

/**
 * @author koqizhao
 *
 * Apr 1, 2018
 */
public class ValidationDecorator<T> implements QuantileEstimator<T> {

    private QuantileEstimator<T> _quantileEstimator;

    public ValidationDecorator(QuantileEstimator<T> quantileEstimator) {
        Objects.requireNonNull(quantileEstimator, "quantileEstimator is null");
        _quantileEstimator = quantileEstimator;
    }

    @Override
    public void add(T value) {
        if (value == null)
            return;

        _quantileEstimator.add(value);
    }

    @Override
    public Map<Double, T> get(List<Double> quantiles) {
        Objects.requireNonNull(quantiles, "quantiles is null");

        if (quantiles.isEmpty())
            return new HashMap<>();

        quantiles.forEach(quantile -> {
            if (quantile == null || quantile < 0 || quantile > 1)
                throw new IllegalArgumentException("quantile is not valid: " + quantile);
        });

        return _quantileEstimator.get(quantiles);
    }

}

File: dotnet/src/MyDotey.QuantileEstimator/Validation/ValidationDecorator.cs
using System;
using System.Collections.Generic;

/**
 * @author koqizhao
 *
 * Apr 1, 2018
 */
namespace MyDotey.Quantile.Validation
{
    public class ValidationDecorator<T> : IQuantileEstimator<T>
    {
        private IQuantileEstimator<T> _quantileEstimator;

        public ValidationDecorator(IQuantileEstimator<T> quantileEstimator)
        {
            if (quantileEstimator == null)
                throw new ArgumentNullException("quantileEstimator is null");

            _quantileEstimator = quantileEstimator;
        }

        public virtual void Add(T value)
        {
            if (value == null)
                return;

            _quantileEstimator.Add(value);
        }

        public virtual Dictionary<Double, T> Get(List<Double> quantiles)
        {
            if (quantiles == null)
                throw new ArgumentNullException("quantiles is null");

            if (quantiles.Count == 0)
                return new Dictionary<Double, T>();

            quantiles.ForEach(quantile =>
            {
                if (quantile < 0 || quantile > 1)
                    throw new ArgumentException("quantile is not valid: " + quantile);
            });

            return _quantileEstimator.Get(quantiles);
        }
    }
}